doc/man/pycardsyncer.1
'\" t
.\"     Title: pycardsyncer
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 11/25/2013
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "PYCARDSYNCER" "1" "11/25/2013" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pycardsyncer \- synchronizes CardDAV resources to local db
.SH "SYNOPSIS"
.sp
\fBpycardsyncer\fR [\fIOPTIONS\fR]
.SH "DESCRIPTION"
.sp
pycardsyncer(1) synchronizes remote CardDAV resources to a local database that can then be queried and modified with pc_query(1)\&.
.SH "OPTIONS"
.PP
\fB\-a, \-\-account\fR=\fINAME\fR
.RS 4
Sync only the account/resource
\fINAME\fR
(this option can be used more than once)\&.
.RE
.PP
\fB\-c, \-\-config\fR=\fICONFIG\fR
.RS 4
Use the configuration file
\fICONFIG\fR, otherwise pycardsyncer will look in
\fI$HOME/\&.pycard/\fR
and
\fI$HOME/\&.config/pycard/\fR
for files named
\fIpycard\&.conf\fR\&.
.RE
.PP
\fB\-\-debug\fR
.RS 4
This option enables debugging output\&.
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
Print a small help text and exit\&.
.RE
.PP
\fB\-v, \-\-version\fR
.RS 4
Print pycardsyncer\(cqs version number and exit\&.
.RE
.SH "AUTHOR"
.sp
pycardsyncer was mostly written by Christian Geier, with a lot of help by others, see \fICONTRIBUTORS\&.txt\fR in the pycarddav distribution\&.
.SH "RESOURCES"
.sp
Main web site: http://lostpackets\&.de/pycarddav/ Please report bugs via the contact information at the above web site or via github: http://github\&.com/geier/khal/\&.
.SH "SEE ALSO"
.sp
pc_query(1), pycard\-import(1)
.SH "COPYING"
.sp
Copyright (C) 2011\-2013 Christian Geier and Contributors\&. pycardsyncer and pycarddav are released under the terms of the Expat/MIT license, see the \fICOPYING\fR file distributed with pycarddav\&.


NEWS
News
====

15.06.2013: pyCardDAV v.0.5.0 released

 **NEW** This release brings support for multiple CardDAV accounts. See the
 usage instructions, the supplied example config and/or pc_query --help. Also
 support for more CardDAV servers is included. If you are upgrading, you need to
 delete the local database, otherwise pyCardDAV will refuse to work.

28.03.2013: pyCardDAV v0.4.1 released

  pyCardDAV v0.4.1 is released. This is a bugfix release, if everything works
  fine for you, there is no need to upgrade. Database deletion should not be
  necessary.

15.11.2012: pyCardDAV v0.4 released

  pyCardDAV v0.4 is released. This is a mayor rewrite (again), so some previously
  fixed bugs might be back in. If your upgrading, you should delete your database
  file first.

  On the plus side, there are some new features in pyCardDAV:

    experimental write support in the backend
    import & export vCards
    import addresses directly from mutt
    speed increase in (initial) sync due to switching from pycurl to requests
    detects removed cards on server and delete them locally
    can delete cards locally and then on server

  Also the license has changed to MIT/Expat, see the COPYING file for details
  (but a beer is still appreciated).

  PyCurl is not required anymore, pyCardDAV relies on requests now (which needs
  to be installed).

  Special thanks to David Soulayrol who made a lot of this happen.

  Attention: please make sure you have a backup when you enable write support, see
  Usage for more details.

27.01.2012: pyCardDAV v0.3.3_ released:
  **New** sabredav/owncloud support, thanks Davide Gerhard.

  Fixes a bug where properties with no type parameters were not printed.

  This release also fixes a small database bug.

  The config file has a new entry (*davserver*) which you can set to
  either davical or sabredav (depending on your CardDAV server).

  Future:
  The source code has been cleaned up quite a bit (nearly every line of code
  has been touched) and some features have been added. Write support is nearly
  finished in the backend (but will probably not be included in the next
  release yet), but the frontend is still really buggy and a pain to use. If
  you want to have a look, check the repository out at github and check the
  branch *write_support* (but it might me broken).

06.01.2012: pyCardDAV v0.3.2_ released:
  this is a minor bugfix update, db deleting should not be necessary. If
  everything is working fine at the moment, there is no need to upgrade.

06.01.2012: pyCardDAV v0.3.1_ released:
  this bugfix release fixes some bugs on Debian and a formatting bug (thanks to
  Antoine Sirinelli) and one more unicode bug (thanks to Thomas Klausner). Also,
  some more meaningful error messages were added.

  **Attention** if you are upgrading:
  you should delete the old database again and resync using pycardsyncer

08.12.2011: pyCardDAV v0.3_ released:
  this fixes an unicode bug and has a lot of internal changes

  **Attention** pc-query has been renamed to pc_query,
  make sure to delete the old database, also the config file
  format has somewhat changed

10.10.2011: pyCardDAV v0.2.1_ released
  this fixes a minor bug in the example config file

14.09.2011: pyCardDAV v0.2_ released
  **New** config files are now supported

13.09.2011: pyCardDAV moved to github_
  feel free to fork etc.

12.08.2011: pyCardDAV v0.1_ released
  first public version


pycarddav/pycardsyncer.py
#!/usr/bin/env python
# coding: utf-8
# vim: set ts=4 sw=4 expandtab sts=4:
# ----------------------------------------------------------------------------
# "THE BEER-WARE LICENSE" (Revision 42):
# <geier@lostpackets.de> wrote this file. As long as you retain this notice you
# can do whatever you want with this stuff. If we meet some day, and you think
# this stuff is worth it, you can buy me a beer in return Christian Geier
# ----------------------------------------------------------------------------
"""
syncs the remote database to the local db
"""
from __future__ import print_function

#try:
from netrc import netrc
from urlparse import urlsplit
from pycarddav import Configuration, ConfigurationParser
from pycarddav import capture_user_interruption
from pycarddav import pycard
from pycarddav import carddav

import getpass
import logging
import sys

#except ImportError, error:
#    sys.stderr.write(error)
#    sys.exit(1)

LEVELS = {'debug': logging.DEBUG,
          'info': logging.INFO,
          'warning': logging.WARNING,
          'error': logging.ERROR,
          'critical': logging.CRITICAL,
          }


class SyncConfigurationParser(ConfigurationParser):
    """A specialized setup tool for synchronization."""
    def __init__(self):
        ConfigurationParser.__init__(self, "syncs the local db to the CardDAV server")
        self.set_mandatory_options([(Configuration.SECTIONS.DAV, 'resource'),
                                    (Configuration.SECTIONS.DB, 'path')])

    def check(self, conf):
        success = ConfigurationParser.check(self, conf)

        if success and not conf.dav__passwd:
            hostname = urlsplit(conf.dav__resource).hostname
            auths = netrc().authenticators(hostname)
            if auths:
                if not conf.dav__user or auths[0] == conf.dav__user:
                    logging.debug("Read password for user %s on %s in .netrc",
                                  auths[0], hostname)
                    conf.dav__user = auths[0]
                    conf.dav__passwd = auths[2]
                else:
                    logging.error("User %s not found for %s in .netrc",
                                  conf.dav__user, hostname)
                    success = False
            elif conf.dav__user:
                conf.dav__passwd = getpass.getpass(prompt='CardDAV password: ')
            else:
                logging.error("Missing credentials for %s", hostname)
                success = False

        return success


def main():
    """this should probably be seperated from the class definitions"""
    capture_user_interruption()

    # Read configuration.
    parser = SyncConfigurationParser()

    conf = parser.parse()
    if conf is None:
        sys.exit(1)

    if conf.debug:
        conf.dump()

    syncer = carddav.PyCardDAV(conf.dav__resource,
                               user=conf.dav__user,
                               passwd=conf.dav__passwd,
                               write_support=conf.write_support,
                               verify = conf.dav__verify)

    my_dbtool = pycard.PcQuery(conf.sqlite__path, "utf-8", "stricts", conf.debug)

    # sync:
    abook = syncer.get_abook()  # type (abook): dict

    for href, etag in abook.iteritems():
        if my_dbtool.needs_update(href, etag):
            logging.debug("getting %s etag: %s", href, etag)
            vcard = syncer.get_vcard(href)
            my_dbtool.update(vcard, href, etag=etag)

    remote_changed = False
    # for now local changes overwritten by remote changes
    logging.info("looking for locally changed vcards...")
    hrefs = my_dbtool.changed
    for href in hrefs:
        logging.info("trying to update %s", href)
        card = my_dbtool.get_vcard_from_db(href)
        card_string = card.serialize()
        card_string = card_string.replace('###COMMA###', ',')
        logging.debug("%s", my_dbtool.get_etag(href))
        syncer.update_vcard(card_string, href, None)
        my_dbtool.reset_flag(href)
    # uploading
    hrefs = my_dbtool.get_new()
    for href in hrefs:
        logging.info("trying to upload new card %s", href)
        card = my_dbtool.get_vcard_from_db(href)
        (href_new, etag_new) = syncer.upload_new_card(card.vcf)
        my_dbtool.update_href(href, href_new)
        remote_changed = True

    # deleting locally deleted cards on the server
    hrefs_etags = my_dbtool.get_marked_delete()
    for href, etag in hrefs_etags:
        logging.info('trying to delete card %s', href)
        syncer.delete_vcard(href, etag)
        my_dbtool.rm_from_deleted(href)
        remote_changed = True

    # detecting remote-deleted cards
    ulist = list()
    # is there a better way to compare a list of unicode() with a list of str()
    # objects?
    if remote_changed:
        abook = syncer.get_abook()  # type (abook): dict
    for one in abook.keys():
        ulist.append(unicode(one))
    rlist = my_dbtool.get_all_vref_from_db()
    delete = set(rlist).difference(ulist)
    for href in delete:
        my_dbtool.delete_vcard_from_db(href)

if __name__ == "__main__":
    main()
